cppfd/solver.cpp

```diff
@@ -2,41 +2,114 @@
 
 #include"solver.h"
 
-Solver::Solver(Mesh& m, double d_t, double t_f, double r, double d_v, double m_C, int v)
-  : mesh(m)
-  , delta_t(d_t)
-  , t_final(t_f)
-  , rho(r)
-  , max_C(m_C)
-  , verbosity(v)
-  {
-    // compute and store kinematic viscosity
-    this->nu = d_v / r;
+// main function to solve N-S equation on time steps
+void Solver::solve(){
+  // begin simulation
+  if(this->verbosity >= 1){
+    std::cout<<"Initial timestep : " << this->delta_t << ", " << "Time final : " << this->t_final <<std::endl;
   }
 
-// main function to solve N-S equation on time steps
-void Solver::solve(BoundaryConditions& b_c)
-{
-  // implement me
+  Kokkos::Timer timer;
+
+  if(this->verbosity >= 1){
+    std::cout<<"Applying velocity boundary conditions..."<<std::endl;
+  }
+  this->boundary_conditions.apply_velocity_bc();
+
+  double velocity_bc_max = this->boundary_conditions.get_velocity_bc_max_norm();
+  double l = this->mesh.get_cell_size() * std::max(this->mesh.get_n_cells_x(), this->mesh.get_n_cells_y());
+  this->Re = velocity_bc_max * l / this->nu;
+  if(this->verbosity > 0){
+    std::cout << "Reynolds number : " << this->Re <<std::endl;
+  }
+
+  double time1 = timer.seconds();
+  if(this->verbosity > 0){
+    std::cout<<"Computing laplacian matrix..."<<std::endl;
+  }
+
+  this->assemble_laplacian();
+  double time2 = timer.seconds();
+  if(this->verbosity > 0){
+    std::cout<<"Laplacian computation duration: " << time2 - time1 <<std::endl;
+  }
 
-  b_c.apply_velocity_bc();
+  if(this->verbosity > 0){
+    std::cout<<"Running simulation..."<<std::endl;
+  }
+
+  double t = 0;
+  int iteration = 0;
+
+  //start iterating on time steps
+  while(t < this->t_final){
+    ++iteration;
+    t += this->delta_t;
+
+    if(this->verbosity >= 2){
+      std::cout<<"Entering iteration " << iteration << " at time " << t <<std::endl;
+    }
+  }
+
+  this->boundary_conditions.apply_velocity_bc();
   this->predict_velocity();
 }
 
+// computation of laplacian matrix
+void Solver::assemble_laplacian(){
+  // initialize container for Laplacian
+  const int m = this->mesh.get_n_cells_x();
+  const int n = this->mesh.get_n_cells_y();
+  const int mn = m * n;
+  this->laplacian = Kokkos::View<double**>("laplacian", mn, mn);
+
+  for(int j = 1; j <= n - 1; j++){
+    for(int i = 1; i <= m - 1; i++){
+      int k = this->mesh.cartesian_to_index(i, j, m, n);
```

Comment:
Add comment here to indicate that the flat index of a cell into the Cartesian mesh becomes the per-dimension Cartesian index into the Laplacian matrix.